felix/nftables/table.go

```diff
@@ -0,0 +1,1293 @@
+// fixes in the iptables/iptables-save command.
+//
+// (5) Support for graceful restart.  I.e. deferring potentially incorrect updates until we're
+// in-sync with the datastore.  For example, if we have 100 endpoints on a host, after a restart
+// we don't want to write a "dispatch" chain when we learn about the first endpoint (possibly
+// replacing an existing one that had all 100 endpoints in place and causing traffic to glitch);
+// instead, we want to defer until we've seen all 100 and then do the write.
+//
+// (6) Improved handling of rule inserts vs Felix 1.4.x.  Previous versions of Felix sometimes
+// inserted special-case rules that were not marked as Calico rules in any sensible way making
+// cleanup of those rules after an upgrade difficult.
+//
+// # Implementation
+//
+// For high performance (goal 1), we use iptables-restore to do bulk updates to iptables.  This is
+// much faster than individual iptables calls.
+//
+// To allow us to restore rules after they are clobbered by another process (goal 2), we cache
+// them at this layer.  This means that we don't need a mechanism to ask the other layers of Felix
+// to do a resync.  Note: nftablesTable doesn't start a thread of its own so it relies on the main event
+// loop to trigger any dataplane resync polls.
+//
+// There is tension between goals 3 and 4.  In order to avoid full rewrites (goal 3), we need to
+// know what rules are in place, but we also don't want to parse them to find out (goal 4)!  As
+// a compromise, we deterministically calculate an ID for each rule and store it in an iptables
+// comment.  Then, when we want to know what rules are in place, we _do_ parse the output from
+// iptables-save, but only to read back the rule IDs.  That limits the amount of parsing we need
+// to do and keeps it manageable/robust.
+//
+// To support graceful restart (goal 5), we defer updates to the dataplane until Apply() is called,
+// then we do an atomic update using iptables-restore.  As long as the first Apply() call is
+// after we're in sync, the dataplane won't be touched until the right time.  Felix 1.4.x had a
+// more complex mechanism to support partial updates during the graceful restart period but
+// Felix 2.0.0 resyncs so quickly that the added complexity is not justified.
+//
+// To make it easier to manage rule insertions (goal 6), we add rule IDs to those too.  With
+// rule IDs in place, we can easily distinguish Calico rules from non-Calico rules without needing
+// to know exactly which rules to expect.  To deal with cleanup after upgrade from older versions
+// that did not write rule IDs, we support special-case regexes to detect our old rules.
+//
+// # Thread safety
+//
+// nftablesTable doesn't do any internal synchronization, its methods should only be called from one
+// thread.  To avoid conflicts in the dataplane itself, there should only be one instance of
+// nftablesTable for each table in an application.
+type nftablesTable struct {
```

Comment:
This is also a big copy / paste of the iptables implementation. I have cut out some bits that I think aren't needed / relevant for nftables, and adjusted the logic in a few places but it still needs me to go over it with a fine comb.